### PR TITLE
CMake build, allowing parent build

### PR DESCRIPTION
**Pull request by ProExpertProg**
**This is an extension of #17 but with logic to allow building inside the vLLM build (#17 only supports standalone build).**

Added a CMake-based build system to enable building this package from source in the vLLM build. That will allow using PyTorch nightly, which is needed by the torch.compile and another project in vLLM.

### CODE REVIEW

File: tests/test_vllm_flash_attn.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        causal=True,
+        block_table=block_tables,
+        cache_seqlens=kv_lens_tensor,
+        softcap=soft_cap if soft_cap is not None else 0,
+    ).squeeze(1)
+
+    if num_blocks <= 2048:
+        test_utils = ["test_faketensor", "test_schema"]
+    else:
+        test_utils = ["test_faketensor"]
```
Comment (bnellnm) [inline tests/test_vllm_flash_attn.py:127]: Can you add a comment about why `test_schema` is disabled for num_blocks > 2048?

File: cmake/utils.cmake [hunk 2]
```diff
@@ -0,0 +1,369 @@
+# This file is taken from github.com/vllm-project/vllm/cmake/utils.cmake
+# It contains utility functions for building PyTorch extensions with GPU support.
+# The only modification is finding the `Python` package with a specific version.
+#
+# THIS MACRO WAS MODIFIED
+# Attempt to find the python package and verifies that it is one of the `SUPPORTED_VERSIONS`.
+# To customize which python gets found, set the `Python_EXECUTABLE` variable.
+macro (find_python_constrained_versions SUPPORTED_VERSIONS)
```
Comment (bnellnm) [inline cmake/utils.cmake:8]: I forgot to comment about this on #17 but the old process for finding the python version based off the executable was very intentional.  Some of the CI systems had multiple versions of python installed and it was important to select the specific version requested rather than just the first in the list of supported verions.

**Comment by bnellnm** [comment]
Looks ok to me except for the python finding logic.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+#
+# Currently, having an incorrect pytorch version results in a warning
+# rather than an error.
+#
+# Note: the CUDA torch version is derived from pyproject.toml and various
+# requirements.txt files and should be kept consistent.  The ROCm torch
+# versions are derived from Dockerfile.rocm
+#
+set(TORCH_SUPPORTED_VERSION_CUDA "2.4.0")
+set(TORCH_SUPPORTED_VERSION_ROCM "2.5.0")
```
Comment (tlrmchlsmth) [inline CMakeLists.txt:38]: I see this is inherited from vLLM's CMakeLists.txt -- anyone know why?

Related: Any ideas on how to keep this kind of thing in sync between vLLM and vllm-flash-attn?

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+list(APPEND VLLM_FA_GPU_FLAGS --expt-relaxed-constexpr --expt-extended-lambda --use_fast_math)
+
+# Replace instead of appending, nvcc doesn't like duplicate -O flags.
+string(REPLACE "-O2" "-O3" CMAKE_CUDA_FLAGS_RELWITHDEBINFO "${CMAKE_CUDA_FLAGS_RELWITHDEBINFO}")
+
+#
+# _C extension
+#
+
+# TODO(luka): happy to hardcode, but these are generated so globbing seems fine.
```
Comment (tlrmchlsmth) [inline CMakeLists.txt:167]: IMO globbing is fine (nit: and remove this comment before landing)

**Review by tlrmchlsmth — APPROVED** [review]
LGTM!

I think you told me this, just wanted to doublecheck: You removed those tests because they are testing code from the original flash-attn repo that was removed as well, right?

File: cmake/hipify.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
```
Comment (WoosukKwon) [inline cmake/hipify.py:1]: QQ: Why do we need hipify?

IIUC, the current version of flash-attn does not support ROCm.

File: csrc/flash_attn/flash_api.cpp [hunk 6]
```diff
@@ -259,14 +260,14 @@ std::vector<at::Tensor>
...
+        const c10::optional<at::Tensor> &alibi_slopes_, // num_heads or batch_size x num_heads
+        const double p_dropout,
+        const double softmax_scale,
         bool is_causal,
-        int window_size_left,
-        int window_size_right,
-        const float softcap,
+        int64_t window_size_left,
+        int64_t window_size_right,
+        const double softcap,
```
Comment (WoosukKwon) [inline csrc/flash_attn/flash_api.cpp:270]: Just curious: how are these type changes related to this PR? Does CMake have any restriction on that?

File: tests/test_flash_attn.py
Comment (WoosukKwon) [inline tests/test_flash_attn.py:1]: Maybe we can just delete this file as well?

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+)
+
+target_include_directories(vllm_flash_attn_c PRIVATE
+        csrc/flash_attn
+        csrc/flash_attn/src
+        csrc/cutlass/include)
+
+# custom definitions
+target_compile_definitions(vllm_flash_attn_c PRIVATE FLASHATTENTION_DISABLE_DROPOUT FLASHATTENTION_DISABLE_UNEVEN_K)
+
```
Comment (WoosukKwon) [inline CMakeLists.txt:189]: Can you comment other variables like `FLASHATTENTION_DISABLE_ALIBI` so that we can easily track those?

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for this huge work! Left some minor comments.

File: csrc/flash_attn/flash_api.cpp (hunk 6 above)
Comment (ProExpertProg) [reply csrc/flash_attn/flash_api.cpp:270]: These are `torch_extension` restrictions iirc

File: cmake/hipify.py (hunk 5 above)
Comment (ProExpertProg) [reply cmake/hipify.py:1]: This was here initially to make it easier to keep the CMake build system (especially `utils.cmake`) as similar to the vLLM repo as possible. But I'll delete this file

File: CMakeLists.txt (hunk 3 above)
Comment (ProExpertProg) [reply CMakeLists.txt:38]: I am not really sure about the answers to any of those three questions 😅 mind if I leave as is and we revisit these? I''l clean up the comment though

**Comment by ProExpertProg** [comment]
All comments addressed - should be good to merge apart from the Torch version in CMake comment @tlrmchlsmth

File: CMakeLists.txt (hunk 3 above)
Comment (tlrmchlsmth) [reply CMakeLists.txt]: Totally 😁
